File: leetcode/python/000131.palindrome-partitioning.py
class Solution:
    # Time complexity: O(2^n * n)
    # Space complexity: O(n) - recursion and temp; O(2^n) if including res
    def partition(self, s: str) -> list[list[str]]:
        res, temp = [], []

        def dfs(i):
            if i == len(s):
                res.append(temp.copy())
                return

            for j in range(i, len(s)):
                if s[i:j+1] == s[i:j+1][::-1]: # check if palindrome
                    temp.append(s[i:j+1])
                    dfs(j + 1)
                    temp.pop()

        dfs(0)
        return res

File: leetcode/typescript/000131.palindrome-partitioning.ts
// Time complexity: O(2^n * n)
// Space complexity: O(n)
function partition(s: string): string[][] {
  const res: string[][] = [];
  const temp: string[] = [];

  function dfs(i: number) {
    if (i === s.length) {
      res.push([...temp]);
      return;
    }

    for (let j = i; j < s.length; j++) {
      const part = s.slice(i, j + 1);
      if (isPalindrome(part)) {
        temp.push(part);
        dfs(j + 1);
        temp.pop();
      }
    }
  }

  dfs(0);
  return res;
}

function isPalindrome(s: string) {
  let l = 0;
  let r = s.length - 1;

  while (l <= r) {
    if (s[l] !== s[r]) return false;
    l++;
    r--;
  }

  return true;
}
